src/torchcodec/decoders/_core/AVIOFileLikeContext.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Meta Platforms, Inc. and affiliates.
+// All rights reserved.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "src/torchcodec/decoders/_core/AVIOFileLikeContext.h"
+#include <torch/types.h>
+
+namespace facebook::torchcodec {
+
+AVIOFileLikeContext::AVIOFileLikeContext(py::object fileLike)
+    : fileLike_{UniquePyObject(new py::object(fileLike))} {
+  {
+    // TODO: Is it necessary to acquire the GIL here? Is it maybe even
+    // harmful? At the moment, this is only called from within a pybind
+    // function, and pybind guarantees we have the GIL.
+    py::gil_scoped_acquire gil;
+    TORCH_CHECK(
+        py::hasattr(fileLike, "read"),
+        "File like object must implement a read method.");
+    TORCH_CHECK(
+        py::hasattr(fileLike, "seek"),
+        "File like object must implement a seek method.");
+  }
+  createAVIOContext(&read, &seek, &fileLike_);
+}
+
+int AVIOFileLikeContext::read(void* opaque, uint8_t* buf, int buf_size) {
+  auto fileLike = static_cast<UniquePyObject*>(opaque);
+
+  int num_read = 0;
+  while (num_read < buf_size) {
+    int request = buf_size - num_read;
+    // TODO: It is maybe more efficient to grab the lock once in the
+    // surrounding scope?
+    py::gil_scoped_acquire gil;
+    auto chunk = static_cast<std::string>(
+        static_cast<py::bytes>((*fileLike)->attr("read")(request)));
```

Comment:
Can you help me understand why casting to `string`? Is it for convenience of keeping both the data and the size together?

Reply:
This is code copied from TorchAudio, and I thought it was necessary for that reason. But looking at the docs, I think we can just call a C++ version of `len()` on the returned `py::bytes`. I'll look into that.

Reply:
Looking more: it's so that we have a pointer to the data that we can pass to `memcpy()`.

Reply:
Shouldn't we be able to get both a pointer and the size from `py::bytes`? It might be simpler to keep `string`, but a small comment may help conveying the intent/reason.

Reply:
We actually can't get the pointer. We can get the size from all `py::object` instances, but there's nothing to access the underlying data pointer. (Other than going through the actual `PyObject` struct, but I'd rather stick to the pybind11 API.)

However! Other people have asked the same thing: https://github.com/pybind/pybind11/issues/2517, which lead to someone else implementing something just as good, which is getting a `std::string_view`: https://github.com/pybind/pybind11/issues/1807. That will avoid the copy and get what we need. Good catch!